compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -220,6 +220,91 @@ impl<'tcx> fmt::Display for LayoutError<'tcx> {
     }
 }
 
+/// Enforce some basic invariants on layouts.
+fn sanity_check_layout<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+    layout: &TyAndLayout<'tcx>,
+) {
+    // Type-level uninhabitedness should always imply ABI uninhabitedness.
+    if tcx.conservative_is_privately_uninhabited(param_env.and(layout.ty)) {
+        assert!(layout.abi.is_uninhabited());
+    }
+
+    if cfg!(debug_assertions) {
+        fn check_layout_abi<'tcx>(tcx: TyCtxt<'tcx>, layout: Layout<'tcx>) {
+            match layout.abi() {
+                Abi::Scalar(_scalar) => {
+                    // No padding in scalars.
+                    /* FIXME(#96185):
+                    assert_eq!(
+                        layout.align().abi,
+                        scalar.align(&tcx).abi,
+                        "alignment mismatch between ABI and layout in {layout:#?}"
+                    );
+                    assert_eq!(
+                        layout.size(),
+                        scalar.size(&tcx),
+                        "size mismatch between ABI and layout in {layout:#?}"
+                    );*/
+                }
+                Abi::ScalarPair(scalar1, scalar2) => {
+                    // Sanity-check scalar pair size.
+                    let field2_offset = scalar1.size(&tcx).align_to(scalar2.align(&tcx).abi);
+                    let total = field2_offset + scalar2.size(&tcx);
+                    assert!(
+                        layout.size() >= total,
+                        "size mismatch between ABI and layout in {layout:#?}"
+                    );
+                }
```

Comment:
Hmm, shouldn't this also have an alignment check?

Honestly, reusing `LayoutCx::scalar_pair` would probably make this really straight-forward (and maybe some miri/codegen code as well, tho in that case it should be a `Layout` constructor rather than on `LayoutCx`) - except for the fact that it allocates two `Vec`s (if only they were `SmallVec<[_; 2]>`... - more things that aren't immediately actionable *sigh*)

Reply:
I wasn't sure what the alignment check would be, so this felt like better than nothing. :shrug: 

I don't see how this could reuse `scalar_pair`; things with ScalarPair ABI can have any number of fields since there can be newtypes around the underlying pair.

Reply:
Oh yeah, the fields would need to be ignored wrt validation (but e.g. codegen could use the offset of the second field to avoid computing `a_size.align_to(b_align)` itself).

Anyway, not that viable today with the allocations.